File: sdk/go/view/response_pipeline.go
package view

import (
	"fmt"

	openrtb "github.com/oakrtb/openrtb/sdk/go/oakrtb/v2"
)

// ResponsePipeline 按序编排 LightGate → SharedView → BidViews（BidResponse）。
type ResponsePipeline struct {
	res      *openrtb.BidResponse
	shared   ResponseSharedView
	seatBids []SeatBidView
	bids     []BidView
	gated    bool
	pinned   bool
	viewed   bool
	err      error
}

// OfResponse 对已解码的 BidResponse 启动流水线编排。
func OfResponse(res *openrtb.BidResponse) *ResponsePipeline {
	return &ResponsePipeline{res: res}
}

// RunResponse 执行全部响应步骤并返回 ResponseSnapshot。
func RunResponse(res *openrtb.BidResponse) (*ResponseSnapshot, error) {
	return OfResponse(res).LightGate().Shared().Bids().Snapshot()
}

// LightGate 为第 1 步：轻量结构检查。
func (p *ResponsePipeline) LightGate() *ResponsePipeline {
	if p.err != nil {
		return p
	}
	if err := LightGateResponse(p.res); err != nil {
		p.err = err
		return p
	}
	p.gated = true
	return p
}

// Shared 为第 2 步：构建拍卖级 ResponseSharedView。
func (p *ResponsePipeline) Shared() *ResponsePipeline {
	if p.err != nil {
		return p
	}
	if p.pinned {
		return p
	}
	if !p.gated {
		p.err = fmt.Errorf("pipeline: call lightGate first")
		return p
	}
	p.shared = responseShared(p.res)
	p.pinned = true
	return p
}

// Bids 为第 3 步：构建 SeatBidView 与扁平 BidView 列表。
func (p *ResponsePipeline) Bids() *ResponsePipeline {
	if p.err != nil {
		return p
	}
	if !p.pinned {
		p.err = fmt.Errorf("pipeline: call shared first")
		return p
	}
	p.seatBids, p.bids = viewSeatBids(p.res)
	p.viewed = true
	return p
}

// Snapshot 完成流水线（自动按序补跑未执行的步骤）。
func (p *ResponsePipeline) Snapshot() (*ResponseSnapshot, error) {
	if p.err != nil {
		return nil, p.err
	}
	if !p.gated {
		p.LightGate()
		if p.err != nil {
			return nil, p.err
		}
	}
	if !p.pinned {
		p.Shared()
		if p.err != nil {
			return nil, p.err
		}
	}
	if !p.viewed {
		p.Bids()
		if p.err != nil {
			return nil, p.err
		}
	}
	return &ResponseSnapshot{Shared: p.shared, SeatBids: p.seatBids, Bids: p.bids}, nil
}

// ResponseSnapshot 是已完成响应流水线的可查询输出。
type ResponseSnapshot struct {
	Shared   ResponseSharedView
	SeatBids []SeatBidView
	Bids     []BidView
}

// RequestID 返回对应请求的 id（BidResponse.id）。
func (s *ResponseSnapshot) RequestID() string { return s.Shared.ID }

// BidID 返回买方响应 id（BidResponse.bidid）。
func (s *ResponseSnapshot) BidID() string { return s.Shared.BidID }

// Currency 返回响应货币（BidResponse.cur）。
func (s *ResponseSnapshot) Currency() string { return s.Shared.Cur }

// NoBid 报告是否为 no-bid（无 seatbid）。
func (s *ResponseSnapshot) NoBid() bool { return s.Shared.NoBid }

// Nbr 返回结构化不出价原因码（BidResponse.nbr）。
func (s *ResponseSnapshot) Nbr() int32 { return s.Shared.Nbr }

// Response 返回底层 BidResponse 指针。
func (s *ResponseSnapshot) Response() *openrtb.BidResponse { return s.Shared.Res }

// FindBid 按 bid id 查找 BidView，未找到时返回 nil。
func (s *ResponseSnapshot) FindBid(id string) *BidView {
	for i := range s.Bids {
		if s.Bids[i].ID == id {
			return &s.Bids[i]
		}
	}
	return nil
}

// BidsForImp 返回指定 impid 的全部 BidView。
func (s *ResponseSnapshot) BidsForImp(impid string) []BidView {
	var out []BidView
	for _, b := range s.Bids {
		if b.ImpID == impid {
			out = append(out, b)
		}
	}
	return out
}

// BidsWith 返回包含指定 markup 标志的 BidView 列表。
func (s *ResponseSnapshot) BidsWith(flag MarkupMask) []BidView {
	var out []BidView
	for _, b := range s.Bids {
		if b.Markup.Has(flag) {
			out = append(out, b)
		}
	}
	return out
}

// Facts 返回用于下游处理的紧凑 per-bid 摘要。
func (s *ResponseSnapshot) Facts() []BidFact {
	out := make([]BidFact, 0, len(s.Bids))
	for _, b := range s.Bids {
		out = append(out, bidFactFrom(b))
	}
	return out
}

// BidFact 是从 BidView 提取的扁平摘要，供调用方使用。
type BidFact struct {
	ID      string
	ImpID   string
	Seat    string
	Price   float64
	Markup  MarkupMask
	Mtype   int32
	Crid    string
	DealID  string
	W       int32
	H       int32
	Dur     int32
	HasAdm  bool
	Adomain []string
}

func bidFactFrom(b BidView) BidFact {
	return BidFact{
		ID:      b.ID,
		ImpID:   b.ImpID,
		Seat:    b.Seat,
		Price:   b.Price,
		Markup:  b.Markup,
		Mtype:   b.Mtype,
		Crid:    b.Crid,
		DealID:  b.DealID,
		W:       b.W,
		H:       b.H,
		Dur:     b.Dur,
		HasAdm:  b.Adm != "",
		Adomain: b.Adomain,
	}
}

// HasBanner 报告 BidFact 是否为 Banner markup。
func (f BidFact) HasBanner() bool { return f.Markup.HasBanner() }

// HasVideo 报告 BidFact 是否为 Video markup。
func (f BidFact) HasVideo() bool { return f.Markup.HasVideo() }

// HasAudio 报告 BidFact 是否为 Audio markup。
func (f BidFact) HasAudio() bool { return f.Markup.HasAudio() }

// HasNative 报告 BidFact 是否为 Native markup。
func (f BidFact) HasNative() bool { return f.Markup.HasNative() }

File: sdk/java/src/main/java/com/oakrtb/sdk/view/ResponsePipeline.java
package com.oakrtb.sdk.view;

import com.oakrtb.openrtb.v2.BidResponse;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * 按固定顺序编排 BidResponse view 流程，并暴露 {@link Snapshot}。
 *
 * <ol>
 *   <li>{@link #lightGate}
 *   <li>{@link #shared}
 *   <li>{@link #bids}
 * </ol>
 */
public final class ResponsePipeline {
  private final BidResponse res;
  private ResponseViews.SharedView shared;
  private List<ResponseViews.SeatBidView> seatBids = List.of();
  private List<ResponseViews.BidView> bids = List.of();
  private boolean gated;
  private boolean pinned;
  private boolean viewed;

  private ResponsePipeline(BidResponse res) {
    this.res = Objects.requireNonNull(res, "BidResponse");
  }

  /**
   * 对已解码的 BidResponse 开始编排。
   *
   * @param res BidResponse
   * @return 管道实例
   */
  public static ResponsePipeline of(BidResponse res) {
    return new ResponsePipeline(res);
  }

  /**
   * 按序执行全部步骤并返回快照。
   *
   * @param res BidResponse
   * @return Snapshot
   */
  public static Snapshot run(BidResponse res) {
    return of(res).lightGate().shared().bids().snapshot();
  }

  /**
   * 步骤 1：结构校验。
   *
   * @return 当前管道
   */
  public ResponsePipeline lightGate() {
    ResponseViews.lightGate(res);
    gated = true;
    return this;
  }

  /**
   * 步骤 2：固定 SharedView（须先 lightGate）。
   *
   * @return 当前管道
   */
  public ResponsePipeline shared() {
    requireGated();
    if (pinned) {
      return this;
    }
    shared = ResponseViews.shared(res);
    pinned = true;
    return this;
  }

  /**
   * 步骤 3：生成 SeatBid/Bid 视图（须先 shared）。
   *
   * @return 当前管道
   */
  public ResponsePipeline bids() {
    requirePinned();
    ResponseViews.BidLists lists = ResponseViews.bidLists(res);
    seatBids = lists.seatBids();
    bids = lists.bids();
    viewed = true;
    return this;
  }

  /**
   * 完成管道；跳过步骤将自动补跑。
   *
   * @return Snapshot
   */
  public Snapshot snapshot() {
    if (!gated) {
      lightGate();
    }
    if (!pinned) {
      shared();
    }
    if (!viewed) {
      bids();
    }
    return new Snapshot(shared, seatBids, bids);
  }

  private void requireGated() {
    if (!gated) {
      throw new IllegalStateException("pipeline: call lightGate() first");
    }
  }

  private void requirePinned() {
    requireGated();
    if (!pinned) {
      throw new IllegalStateException("pipeline: call shared() first");
    }
  }

  /**
   * 响应 view 完成后的可查询快照。
   *
   * @param shared 响应级视图
   * @param seatBids SeatBid 视图列表
   * @param bids 扁平 Bid 列表
   */
  public record Snapshot(
      ResponseViews.SharedView shared,
      List<ResponseViews.SeatBidView> seatBids,
      List<ResponseViews.BidView> bids) {

    public Snapshot {
      Objects.requireNonNull(shared, "shared");
      seatBids = List.copyOf(Objects.requireNonNull(seatBids, "seatBids"));
      bids = List.copyOf(Objects.requireNonNull(bids, "bids"));
    }

    /** 回显的请求 id。 */
    public String requestId() {
      return shared.id();
    }

    /** 响应级 bid id。 */
    public String bidid() {
      return shared.bidid();
    }

    /** 响应货币。 */
    public String currency() {
      return shared.cur();
    }

    /** 是否为 wire no-bid（无 seatbid）。 */
    public boolean noBid() {
      return shared.noBid();
    }

    /** no-bid 原因码。 */
    public int nbr() {
      return shared.nbr();
    }

    /** 原始 BidResponse。 */
    public BidResponse response() {
      return shared.res();
    }

    /**
     * 按 bid id 查找。
     *
     * @param id bid id
     * @return BidView，无则 empty
     */
    public Optional<ResponseViews.BidView> findBid(String id) {
      if (id == null) {
        return Optional.empty();
      }
      return bids.stream().filter(b -> id.equals(b.id())).findFirst();
    }

    /**
     * 按 impid 筛选出价。
     *
     * @param impid Imp id
     * @return 匹配的 Bid 列表
     */
    public List<ResponseViews.BidView> bidsForImp(String impid) {
      List<ResponseViews.BidView> out = new ArrayList<>();
      for (ResponseViews.BidView b : bids) {
        if (Objects.equals(impid, b.impid())) {
          out.add(b);
        }
      }
      return Collections.unmodifiableList(out);
    }

    /**
     * 筛选包含指定格式位的出价。
     *
     * @param formatFlag {@link MarkupMask} 单 bit 常量
     * @return 匹配的 Bid 列表
     */
    public List<ResponseViews.BidView> bidsWith(int formatFlag) {
      List<ResponseViews.BidView> out = new ArrayList<>();
      for (ResponseViews.BidView b : bids) {
        if (b.markup().has(formatFlag)) {
          out.add(b);
        }
      }
      return Collections.unmodifiableList(out);
    }

    /**
     * 提取各 Bid 的紧凑事实。
     *
     * @return BidFact 列表
     */
    public List<BidFact> facts() {
      List<BidFact> out = new ArrayList<>(bids.size());
      for (ResponseViews.BidView b : bids) {
        out.add(BidFact.from(b));
      }
      return Collections.unmodifiableList(out);
    }
  }

  /**
   * 单条 Bid 的扁平事实。
   *
   * @param id bid id
   * @param impid Imp id
   * @param seat 席位
   * @param price 价格
   * @param markup 格式掩码
   * @param mtype mtype 数值
   * @param crid 创意 id
   * @param dealid deal id
   * @param w 宽
   * @param h 高
   * @param dur 时长
   * @param hasAdm 是否有 adm
   * @param adomain 广告主域名
   */
  public record BidFact(
      String id,
      String impid,
      String seat,
      double price,
      MarkupMask markup,
      int mtype,
      String crid,
      String dealid,
      int w,
      int h,
      int dur,
      boolean hasAdm,
      List<String> adomain) {

    static BidFact from(ResponseViews.BidView b) {
      return new BidFact(
          b.id(),
          b.impid(),
          b.seat(),
          b.price(),
          b.markup(),
          b.mtype(),
          b.crid(),
          b.dealid(),
          b.w(),
          b.h(),
          b.dur(),
          b.adm() != null && !b.adm().isBlank(),
          b.adomain());
    }

    /** 是否 Banner 出价。 */
    public boolean hasBanner() {
      return markup.hasBanner();
    }

    /** 是否 Video 出价。 */
    public boolean hasVideo() {
      return markup.hasVideo();
    }

    /** 是否 Audio 出价。 */
    public boolean hasAudio() {
      return markup.hasAudio();
    }

    /** 是否 Native 出价。 */
    public boolean hasNative() {
      return markup.hasNative();
    }
  }
}
